src/H5PLplugin_cache.c

```diff
@@ -263,32 +263,89 @@ H5PL__find_plugin_in_cache(const H5PL_search_params_t *search_params, hbool_t *f
 
     /* Loop over all the plugins, looking for one that matches */
     for (u = 0; u < H5PL_num_plugins_g; u++) {
-
-        /* If the plugin type (filter, VOL connector, etc.) and ID match, query the plugin for its info */
-        if ((search_params->type == (H5PL_cache_g[u]).type) &&
-            (search_params->key->id == (H5PL_cache_g[u]).key.id)) {
```

Comment:
Plugin caching only took filter IDs into account and would probably cause an invalid read for VOL/VFD plugins. This should fix caching for both of those.